### PR TITLE
refactor(editor-ui): move json data view into its own component

### PR DESCRIPTION
**Pull request by cstuncsik**
2nd step of implementing JSON mapping: create a separate component for easier development experience and maintenance

### CODE REVIEW

File: packages/editor-ui/src/components/RunData.vue [hunk 1]
```diff
@@ -243,21 +243,11 @@
...
-					path=""
-					:highlightSelectedNode="true"
-					:selectOnClickNode="true"
-					@click="dataItemClicked"
-					class="json-data"
-				/>
-			</div>
+			<run-data-json
+				v-else-if="hasNodeRun && displayMode === 'json'"
+				:jsonData="jsonData"
```
Comment (mutdmour) [inline packages/editor-ui/src/components/RunData.vue:248]: for completeness, you should move jsonData computed function into `run-data-json`, pass in `inputData` as prop similar to table view and make convertToJSON a helper function..

**Review by mutdmour — COMMENTED** [review]
nice :)

File: packages/editor-ui/src/components/Draggable.vue [hunk 2]
```diff
@@ -73,61 +73,64 @@ export default Vue.extend({
...
-				this.draggablePosition = { x: this.stickyPosition[0], y: this.stickyPosition[1]};
-			}
-			else {
-				this.draggablePosition = { x: e.pageX, y: e.pageY };
+				const data = this.targetDataKey && this.draggingEl ? this.draggingEl.dataset.value : (this.data || '');
+				this.$store.commit('ui/draggableStartDragging', {type: this.type, data });
+
+				this.$emit('dragstart', this.draggingEl);
+				document.body.style.cursor = 'grabbing';
+				document.body.classList.add('disable-selection');
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/Draggable.vue]: Actually if block the event only during drag (mousemove) it is enough.
Unfortunately preventing default behaviour and propagation on mousedown is disabled the click on json props and there was no selection performed (on the slot provided by the `vue-json-pretty` component)

Fixed

File: packages/editor-ui/src/components/Draggable.vue [hunk 3]
```diff
@@ -73,61 +73,64 @@ export default Vue.extend({
...
 			}
 
-			this.$emit('drag', this.draggablePosition);
+			this.animationFrameId = window.requestAnimationFrame(() => {
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/Draggable.vue:116]: `requestAnimationFrame` is the best way to keep animations smooth and also spare some CPU/GPU work.
By passing a callback you tell the browser: whenever it fits for your schedule please call this function (preventing it called more than it needs, imagine a how many times a function called during mouse move or scroll)
It's something similar than throttling but relying on browser repaint mechanism.

https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame

File: packages/editor-ui/src/components/RunData.vue [hunk 4]
```diff
@@ -410,13 +380,14 @@ export default mixins(
...
 			NodeErrorView,
-			VueJsonPretty,
 			WarningTooltip,
 			CodeEditor,
 			RunDataTable,
+			RunDataJson,
 		},
 		props: {
 			nodeUi: {
+				type: Object as () => INodeUi,
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunData.vue]: done

File: packages/editor-ui/src/components/RunDataJsonActions.vue [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+// A path that does not exist so that nothing is selected by default
+const nonExistingJsonPath = '_!^&*';
+
+export default mixins(
+	genericHelpers,
+	nodeHelpers,
+	pinData,
+	copyPaste,
+).extend({
+	name: 'run-data-json-actions',
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJsonActions.vue:53]: thanks

File: packages/editor-ui/src/components/RunDataJson.vue [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+						data-target="mappable"
+						:data-value="getJsonParameterPath(node.path)"
+						:data-name="node.key"
+						:data-path="node.path"
+						:class="{
+							[$style.mappable]: mappingEnabled,
+							[$style.dragged]: draggingPath === node.path,
+						}"
+					>
+						{{ node.key }}
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue]: done

File: packages/editor-ui/src/components/RunDataJson.vue [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+			v-if="!editMode.enabled"
+			:node="node"
+			:sessioId="sessionId"
+			:displayMode="displayMode"
+			:distanceFromActive="distanceFromActive"
+			:selectedJsonPath="selectedJsonPath"
+			:jsonData="jsonData"
+			:paneType="paneType"
+		/>
+		<Draggable
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue:13]: do you mean the passed prop names on the child component?

File: packages/editor-ui/src/components/RunData.vue [hunk 8]
```diff
@@ -239,28 +211,32 @@
...
+				v-else-if="hasNodeRun && displayMode === 'json'"
+				:paneType="paneType"
+				:editMode="editMode"
+				:currentOutputIndex="currentOutputIndex"
+				:sessioId="sessionId"
+				:node="node"
+				:inputData="inputData"
+				:mappingEnabled="mappingEnabled"
+				:distanceFromActive="distanceFromActive"
+				:showMappingHint="showMappingHint"
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunData.vue:236]: yes, next round, added to the description of the tooltip task

File: packages/editor-ui/src/components/RunDataJson.vue [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+			}, 6000);
+		}
+
+		if (this.showMappingHint && this.showHint) {
+			setTimeout(() => {
+				this.showHintWithDelay = this.showHint;
+				this.$telemetry.track('User viewed JSON mapping tooltip', { type: 'param focus' });
+			}, 500);
+		}
+	},
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue:137]: yes, same as for `showMappinghint `, will be handled in the tooltip task

File: packages/editor-ui/src/components/RunDataJson.vue [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+					:showLength="true"
+					:selected-value.sync="selectedJsonPath"
+					rootPath=""
+					selectableType="single"
+					class="json-data"
+				>
+					<template #nodeKey="{ node }">
+					<span
+						data-target="mappable"
+						:data-value="getJsonParameterPath(node.path)"
```
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue:39]: can you give an example?

File: packages/editor-ui/src/components/RunDataJson.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/RunDataJson.vue:13]: the component name

File: packages/editor-ui/src/components/RunDataJson.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (mutdmour) [reply packages/editor-ui/src/components/RunDataJson.vue:39]: yes try keys with `.` or space or quotes... you need to call convertPath which is not getting called https://github.com/n8n-io/n8n/pull/4158/files#diff-a40463d9a7bb08a4f2c04ce8d65293ee9059db17fb906dcf6f72567060851eabR136

File: packages/editor-ui/src/components/RunDataJson.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue]: done

File: packages/editor-ui/src/components/RunDataJson.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (cstuncsik) [reply packages/editor-ui/src/components/RunDataJson.vue:39]: done

**Pull request merged**